application/src/main/java/org/thingsboard/server/service/gateway_device/DefaultGatewayDeviceStateService.java

```diff
@@ -0,0 +1,106 @@
+    @Lazy
+    @Autowired
+    private TbCoreDeviceRpcService deviceRpcService;
+
+    @Override
+    public void update(Device device, Device oldDevice) {
+        Device gatewayDevice = findGatewayDeviceByAdditionalInfoInDevice(device.getTenantId(), device.getAdditionalInfo());
+        if (gatewayDevice != null) {
+            ObjectNode renamedDeviceNode = JacksonUtil.newObjectNode();
+            renamedDeviceNode.put(oldDevice.getName(), device.getName());
+            ToDeviceRpcRequest rpcRequest = formDeviceToGatewayRPCRequest(gatewayDevice, renamedDeviceNode, DEVICE_RENAMED_METHOD_NAME);
+            deviceRpcService.processRestApiRpcRequest(rpcRequest, fromDeviceRpcResponse -> {
+                log.trace("Device renamed RPC with id: [{}] processed to gateway device with id: [{}], old device name: [{}], new device name: [{}]",
+                        rpcRequest.getId(), gatewayDevice.getId(), oldDevice.getName(), device.getName());
+            }, null);
+        }
+    }
+
+    @Override
+    public void delete(Device device) {
+        Device gatewayDevice = findGatewayDeviceByAdditionalInfoInDevice(device.getTenantId(), device.getAdditionalInfo());
+        if (gatewayDevice != null) {
+            TextNode deletedDeviceNode = new TextNode(device.getName());
+            ToDeviceRpcRequest rpcRequest = formDeviceToGatewayRPCRequest(gatewayDevice, deletedDeviceNode, DEVICE_DELETED_METHOD_NAME);
+            deviceRpcService.processRestApiRpcRequest(rpcRequest, fromDeviceRpcResponse -> {
+                log.trace("Device deleted RPC with id: [{}] processed to gateway device with id: [{}], deleted device name: [{}]",
+                        rpcRequest.getId(), gatewayDevice.getId(), device.getName());
+            }, null);
+        }
+    }
+
+    private ToDeviceRpcRequest formDeviceToGatewayRPCRequest(Device gatewayDevice, JsonNode deviceDataNode, String method) {
+        ToDeviceRpcRequestBody body = new ToDeviceRpcRequestBody(method, JacksonUtil.toString(deviceDataNode));
+        long expTime = System.currentTimeMillis() + Math.max(minTimeout, defaultTimeout);
+        UUID rpcRequestUUID = UUID.randomUUID();
+        return new ToDeviceRpcRequest(rpcRequestUUID,
+                gatewayDevice.getTenantId(),
+                gatewayDevice.getId(),
+                true,
+                expTime,
+                body,
+                true,
+                null,
+                null
+        );
+    }
+
+    private Device findGatewayDeviceByAdditionalInfoInDevice(TenantId tenantId, JsonNode deviceAdditionalInfo) {
+        if (deviceAdditionalInfo != null && deviceAdditionalInfo.has(DataConstants.LAST_CONNECTED_GATEWAY)) {
+            JsonNode lastConnectedGatewayIdNode = deviceAdditionalInfo.get(DataConstants.LAST_CONNECTED_GATEWAY);
+            DeviceId gatewayId = new DeviceId(UUID.fromString(lastConnectedGatewayIdNode.asText()));
+            return deviceService.findDeviceById(tenantId, gatewayId);
```

Comment:
Please, just return the gatewayId. No need to fetch the gateway and check that it exists. We will simply create an RPC call for missing gateway and it will be removed by TTL function.